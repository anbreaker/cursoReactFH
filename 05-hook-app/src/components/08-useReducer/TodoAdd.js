import React from 'react';
import {useForm} from '../../hooks/useForm';

export const TodoAdd = ({handleAddTodo}) => {
  const [{description}, handleInputChange, reset] = useForm({
    description: '',
  });

  const handleSubmit = (event) => {
    event.preventDefault();
    if (description.trim().length <= 0) return;

    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };

    const action = {
      type: 'add',
      payload: newTodo,
    };

    handleAddTodo(newTodo);
    reset();
  };

  return (
    <>
      <h4>Agregar Todo</h4>
      <hr />

      <form onSubmit={handleSubmit}>
        <input
          className="form-control"
          type="text"
          name="description"
          placeholder="Tarea a rercordar..."
          autoComplete="off"
          value={description}
          onChange={handleInputChange}
        />
        <button type="submit" className="btn btn-primary mt-2 btn-block">
          Agregar
        </button>
      </form>
    </>
  );
};
