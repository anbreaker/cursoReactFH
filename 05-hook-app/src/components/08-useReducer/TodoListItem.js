import React from 'react';

export const TodoListItem = ({todo, index, hadleDelete, handleToggle}) => {
  return (
    <li key={todo.id} className="list-group-item">
      <p className={`${todo.done && 'complete'}`} onClick={() => handleToggle(todo.id)}>
        {index + 1}.- {todo.description}
      </p>

      <button
        className="btn btn-danger"
        onClick={() => {
          hadleDelete(todo.id);
        }}>
        Borrar
      </button>
    </li>
  );
};
