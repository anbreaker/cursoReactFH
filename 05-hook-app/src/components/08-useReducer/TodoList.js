import React from 'react';
import {TodoListItem} from './TodoListItem';

export const TodoList = ({todos, hadleDelete, handleToggle}) => {
  return (
    <ul className="list-group list-group-flush">
      {todos.map((todo, index) => (
        <TodoListItem
          key={todo.id}
          todo={todo}
          index={index}
          hadleDelete={hadleDelete}
          handleToggle={handleToggle}
        />
      ))}
    </ul>
  );
};
