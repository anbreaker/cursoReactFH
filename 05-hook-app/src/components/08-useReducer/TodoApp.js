import React, {useEffect, useReducer} from 'react';
import {todoReducer} from './todoReducer';
import {useForm} from '../../hooks/useForm';
import './useReducer.css';
import {TodoList} from './TodoList';

const init = () => {
  return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  const [{description}, handleInputChange, reset] = useForm({
    description: '',
  });

  useEffect(() => {
    localStorage.setItem('todos', JSON.stringify(todos));
  }, [todos]);

  const hadleDelete = (todoId) => {
    const action = {
      type: 'delete',
      payload: todoId,
    };

    dispatch(action);
  };

  const handleToggle = (todoId) => {
    const action = {
      type: 'toggle',
      payload: todoId,
    };

    dispatch(action);
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    if (description.trim().length <= 0) return;

    const newTodo = {
      id: new Date().getTime(),
      description: description,
      done: false,
    };

    const action = {
      type: 'add',
      payload: newTodo,
    };

    dispatch(action);
    reset();
  };

  return (
    <div className="container">
      <h1>TodoApp ({todos.length})</h1>
      <hr />

      <div className="row">
        <div className="col-7">
          {/* TodoList, todos */}
          <TodoList todos={todos} hadleDelete={hadleDelete} handleToggle={handleToggle} />
        </div>
        <div className="col-5">
          <h4>Agregar Todo</h4>
          <hr />

          <form onSubmit={handleSubmit}>
            <input
              className="form-control"
              type="text"
              name="description"
              placeholder="Tarea a rercordar..."
              autoComplete="off"
              value={description}
              onChange={handleInputChange}
            />
            <button type="submit" className="btn btn-primary mt-2 btn-block">
              Agregar
            </button>
          </form>
        </div>
      </div>
    </div>
  );
};
