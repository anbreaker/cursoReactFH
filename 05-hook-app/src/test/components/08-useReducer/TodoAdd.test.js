import {shallow} from 'enzyme';
import {TodoAdd} from '../../../components/08-useReducer/TodoAdd';
import {demoTodos} from '../../fixtures/demoTodos';

describe('Pruebas en <TodoAdd /> ', () => {
  const handleAddTodo = jest.fn();

  const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

  test('debe de mostarse correctamente', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('NO debe llamar handleAddTodo', () => {
    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({preventDefault() {}});

    expect(handleAddTodo).toHaveBeenCalledTimes(0);
  });

  test('debe llamar la funcion handleAddTodo', () => {
    const value = 'Aprender React';

    wrapper.find('input').simulate('change', {
      target: {
        value,
        name: 'description',
      },
    });

    const formSubmit = wrapper.find('form').prop('onSubmit');
    formSubmit({preventDefault() {}});

    expect(handleAddTodo).toHaveBeenCalledTimes(1);
    expect(handleAddTodo).toHaveBeenCalledWith(expect.any(Object)); // {}
    expect(handleAddTodo).toHaveBeenCalledWith({
      id: expect.any(Number),
      description: value,
      done: false,
    });

    expect(wrapper.find('input').prop('value')).toBe('');
  });
});
