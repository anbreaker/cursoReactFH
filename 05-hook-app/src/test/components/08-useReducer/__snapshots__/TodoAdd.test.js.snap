// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd />  debe de mostarse correctamente 1`] = `
<Fragment>
  <h4>
    Agregar Todo
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Tarea a rercordar..."
      type="text"
      value=""
    />
    <button
      className="btn btn-primary mt-2 btn-block"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
